smart-hadoop-support/smart-hadoop/src/main/java/org/smartdata/hdfs/action/CopyFileAction.java

```diff
@@ -220,4 +236,25 @@ private short getReplication(Short defaultReplication) {
         ? srcFileStatus.getReplication()
         : defaultReplication;
   }
+
+  @SuppressWarnings("BusyWait")
+  private void truncateFile(FileSystem fileSystem, Path path, long newLength) throws IOException {
+    if (fileSystem.truncate(path, newLength)) {
+      // return if the file is already truncated
+      return;
+    }
+
+    appendLog("Waiting for the file " + path + " to be truncated");
+    try {
+      // wait for the truncation of the blocks on datanodes
+      while (fileSystem.getFileStatus(path).getLen() != newLength) {
```

Comment:
I checked other waiting code in the project and it needs refactoring rather than using libraries like awaitatility. But we can definitely use them in the future for testing purposes.

Reply:
Could this cycle continue forever for some reason? Should not we limit the number of attempts ?

Reply:
At first I thought about adding some global limit on retries/timeout, but it's quite hard to choose versatile value for it and also HDFS guarantees eventual completion of the truncate operation in case if the cluster is healthy. Otherwise if it is still unhealthy after all retries, `fileSystem.getFileStatus()` will throw an exception.